Paper: Christensen-Prototype to Production Paper (#562)
scipy-conference/scipy_proceedings: papers/scott_christensen/paper.rst

:author: Scott Christensen
:email: scott.d.christensen@usace.army.mil
:institution: US Army Engineer Research and Development Center
:corresponding:

:author: Marvin Brown
:email: marvin.s.brown@usace.army.mil
:institution: US Army Engineer Research and Development Center

-------------------------------------------------------------------------
Prototype to Production: Python Tools for Rapid Web Interface Development
-------------------------------------------------------------------------

.. class:: abstract

   Taking a powerful and complex Python workflow to the next level as an interactive web application can make a huge difference in its utility and ability to communicate results. The enhanced capabilities of open-source Python libraries have made making the jump from Jupyter notebooks to production-ready web applications easier than ever. This talk will demonstrate how prototyping a web application can begin in right in the notebook environment and then easily be deployed as a stand-alone web application with Panel. We then show how Panel apps can be transitioned to fully-featured web app in Tethys Platform.

.. class:: keywords

>    web apps, Panel, Tethys Platform, Holoviz

Reviewer: `Jupyter notebook` can be another relevant keyword

Introduction
------------
Python has become a standard for orchestrating complex scientific workflows, which are commonly used to inform decision makers. However, the utility of those workflows is limited when results are delivered as printed reports that evaluate only a limited number of options and try to communicate complex data with a few static figures. Web applications provide a non-technical decision maker with the ability to gain greater insight from a technical scientific workflow by allowing them to explore more scenarios and by providing interactive visualizations. However, developing customized web applications in a rapidly evolving research environment is challenging. This talk will show how several open-source Python projects can be used to enable rapidly prototyping web-based interfaces in a Jupyter notebook, deploying those interfaces as standalone web-apps, and then transitioning those apps into a production ready platform, and is intended for those that need to deliver technical Python workflows to non-technical decision makers.

> The next section provides some background on the core tools that are leveraged and enhanced to create the rapid prototyping capability. That is followed by the Methods section which will describe how those core tools were enhanced and chained together to create a pipeline of tools for taking a prototype to production. The Results section will describe how those tools were used to create a sample web application to analyze and explore COVID-19 data. Finally, the conclusion will provide some discussion and discuss future work that can continue to enhance how these tools integrate.

Reviewer: ```suggestion
The next section provides some background on the core tools that are leveraged and enhanced to create the rapid prototyping capability. That is followed by the Methods section which describes how those core tools were enhanced and chained together to create a pipeline of tools for taking a prototype to production. The Results section describes how those tools were used to create a sample web application to analyze and explore COVID-19 data. Finally, the conclusion provides some discussion and suggests future work that can continue to enhance the proposed integrated pipeline.
```

Background
----------
The workflow described in this paper leverages several existing open source tools. This section provides a brief background on those tools.

> Jupyter notebooks are a commonly used as a starting point for scientific workflows because it is an open source tool that integrates code, interactive visualizations, and documentation in a sharable and reproducible way. The notebook environment is ideal for quickly iterating and prototyping scripts and visualizations.

Reviewer: ```suggestion
Jupyter notebooks are a commonly used as a starting point for scientific workflows because they are open source tools that integrate code, interactive visualizations, and documentation in a sharable and reproducible way. The notebook environment is ideal for quickly iterating and prototyping scripts and visualizations.
```

The Holoviz tool stack is designed to integrate into the Jupyter notebook environment and work similarly to IPywidgets. While Holoviz interoperates with many different plotting libraries to create interactive visualizations it is closely tied to Bokeh both as a plotting library and backend server. This connection to Bokeh enables visualizations that were created in a notebook environment to be deployed with Bokeh independent of the notebook. This capability is ideal for a more polished and mature workflow that can be presented independent of the code that created it.

Tethys Platform is a Python-based framework for developing scientific web applications that is built on top of Django. While it was originally created to support the needs of water resources web applications it has very flexible and powerful computational and visualization tools that meet the needs of many scientific web applications. The goal of Tethys Platform is to lower the barrier to deploying web applications for scientists and engineers.

Methods
-------
This section describes how existing tools were enhanced and new tools created to enable to prototype to production workflow.

> The Holoviz tool stack while providing many interactive widgets and plots lacked the flexibility in controlling the layout and in integrating other elements into a dashboard. We collaborated with the Holoviz team at Anaconda to add a new library to the Holoviz stack called Panel. Panel allows much more control over how elements are organized. It provides several tools for controlling layout including nested rows and columns, grids, tabs, and widget boxes :ref:`fig1`. Panel continues to be actively developed and had gained wide community support; thus, additional features continue to be added.

Reviewer: ```suggestion
The Holoviz tool stack, while providing many interactive widgets and plots, lacked the flexibility in controlling the layout and in integrating other elements into a dashboard. We collaborated with the Holoviz team at Anaconda to add a new library to the Holoviz stack called Panel. Panel allows much more control over how elements are organized. It provides several tools for controlling layout including nested rows and columns, grids, tabs, and widget boxes (Fig. :ref:`fig1`). Panel continues to be actively developed and had gained wide community support; thus, additional features continue to be added.
```

.. figure:: panel_layouts.png
   :align: center

   This illustrates the various layout tools provided by the Panel library. :label:`fig1`

> Additionally, Panel also provides Pipelines, which supports creating multi-step workflows :ref:`fig2`. Many scientific workflows require various stages of user interaction/input and pipelines allows multiple dashboards and visualizations to be strung together in sequence to support these types of workflows.

Reviewer: ```suggestion
Additionally, Panel also provides Pipelines, which supports creating multi-step workflows (Fig. :ref:`fig2`). Many scientific workflows require various stages of user interaction/input and Pipelines allows multiple dashboards and visualizations to be strung together in sequence to support these types of workflows.
```

> .. figure:: panel_pipeline.png

Reviewer: This figure is not completely clear to me. 
It could be more useful to have two panels, one representing the visualization at Stage 1 and another one at Stage 2.
What is the meaning of` 0.0*0.0 = 0.0` and `0.0^0.0 = 1.0`?

   :align: center

   This demonstrates how multiple stages can be sequenced together using Panel Pipelines. :label:`fig2`

The creation of Panel in the Holoviz tool stack accomplished the goal of supporting very customizable and flexible dashboards that can be pieced together in a notebook environment and then transitioned to an independent web application deployed with Bokeh. However, we wanted to continue to the transition into a production environment. To this end, we enabled the integration of Bokeh with Tethys Platform.

We worked with Anaconda to enable the Bokeh server to be integrated with Django Channels, the web socket backend for Django. This integration allows Bokeh applications (including those created with Panel) to be hosted and embedded in a Django-backed website. We also worked with Aquaveo to add support for Django Channels in Tethys Platform and to enable easy integration of Bokeh and Panel apps.

> By integrating Panel apps into git Tethys, we can then leverage the many tools that Tethys provides including user management and permissions, easy database serialization, and the powerful compute API that supports farming out compute jobs to HTCondor, Dask, and (through a plugin) the DoD research HPC systems. These additional features of Tethys make it an ideal platform for deploying production ready scientific workflows.

Reviewer: What is the DoD research HPC system? Is that accessible by anyone?

The creation of Panel and the integration of Bokeh with Tethys creates a well-established pattern for transitioning from Jupyter notebooks to polished, production web applications in Tethys Platform.

Results
-------

> We have now created several web applications for complex numerical modeling workflows and data analysis. As an example case we will demonstrate how an interactive visualization exploration dashboard for COVID-19 data were created into a web application. As a first step, and as is customary in scientific workflows, we first explored the data using a Jupyter notebook :ref:`fig3`

Reviewer: Could you specify what COVID-19 data is it possible to visualize (e.g., cases, deaths, tests performed, etc.)? Which is the database used as source?

> .. figure:: notebook.png

Reviewer 2: Presumably there are also stages for data-cleaning, summarzing, etc (customary in scientific workflows) before the data is ready for presentation in the panel? These are note shown in the figure. The text in the paragraph above is not clear if these steps are supposed to be part of the notebook. 

If the lack of code preamples is due to difficulty in showing code in a screenshot, consider pasting the coe in as a code block in the .rst file.

Author: @rainwoodman Thank you for the feedback. I appreciate your perspective on how to make our research more interesting and useable by the community. We will work on addressing your suggestions and push some update soon.
